File: ls/ls.c
/*
 * $ gcc -o ls ls.c && ./ls
 */

#include <dirent.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>

void list_files(char *name) {
	DIR *dir = opendir(name);
	if (dir == NULL) {
		fprintf(stderr, "%s: %s\n", name, strerror(errno));
		return;
	}

	struct dirent *entry;
	while (1) {
		entry = readdir(dir);
		if (entry == NULL) {
			break;
		}

		char *file = entry->d_name;
		if (!strcmp(file, ".") || !strcmp(file, "..")) {
			continue;
		}

		printf("%s\n", entry->d_name);
	}

	if (closedir(dir)) {
		fprintf(stderr, "%s: %s\n", name, strerror(errno));
		return;
	}
}

int main(int argc, char *argv[])
{
	if (argc == 1) {
		list_files(".");
	} else {
		int i;
		for (i = 1; i<argc; i++) {
			printf("[%s]\n", argv[i]);
			list_files(argv[i]);
			if (i < (argc - 1)) {
				printf("\n");
			}
		}
	}
	return 0;
}

File: ls/ls.go
// $ go run ls.go
package main

import (
	"fmt"
	"os"
)

func list_files(dir string) {
	d, err := os.Open(dir)
	if err != nil {
		fmt.Printf("%s\n", err.Error())
		return
	}

	defer d.Close()

	files, err := d.Readdir(0)
	if err != nil {
		fmt.Printf("%s\n", err.Error())
		return
	}

	for _, file := range files {
		fmt.Println(file.Name())
	}
}

func main() {
	args := os.Args[1:]
	argc := len(args)

	if argc > 0 {
		for i, dir := range args {
			fmt.Printf("[%s]\n", dir)
			list_files(dir)
			if i < argc-1 {
				fmt.Println()
			}
		}
	} else {
		list_files(".")
	}

}
